openmdao/docs/openmdao_book/features/building_blocks/math_library/math_library.ipynb

```diff
@@ -0,0 +1,601 @@
+    "                                 'vjp': self._compute_partials_vjp}\n",
+    "\n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_jacfwd(self, x):\n",
+    "        deriv_func = jax.jacfwd(self.compute_primal, argnums=[0])\n",
+    "        # Always returns a tuple\n",
+    "        drms_dx, dswitch_dx = deriv_func(x)\n",
+    "        return drms_dx, dswitch_dx\n",
+    "    \n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_jacrev(self, x):\n",
+    "        deriv_func = jax.jacrev(self.compute_primal, argnums=[0])\n",
+    "        # Always returns a tuple\n",
+    "        drms_dx, dswitch_dx = deriv_func(x)\n",
+    "        return drms_dx, dswitch_dx\n",
+    "\n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_jvp(self, x):\n",
+    "        # Note that JVP is a poor choice here, since the jacobian is a row vector!\n",
+    "        \n",
+    "        # Multipling the jacobian by a column vector of ones results in a scalar output,\n",
+    "        # and we are unable to identify the elements in the individual columns of the\n",
+    "        # jacobian.\n",
+    "        \n",
+    "        # Instead, we have to set one element of the tangent to 1 while the rest are zero and\n",
+    "        # evaluate the jvp and extract the value in the corresponding column of the jacobian.\n",
+    "        # We have no choice but to do this one element at a time, so if the size of x is large,\n",
+    "        # this gets prohibitively expensive.\n",
+    "        drms_dx = jnp.zeros_like(x)\n",
+    "        dswitch_dx = jnp.zeros_like(x)\n",
+    "        tangents_x = jnp.zeros_like(x)\n",
+    " \n",
+    "        for i in range(len(x)):\n",
+    "            tangents_x = tangents_x.at[i].set(1.)\n",
+    "            # jvp always returns the primal and the jvp\n",
+    "            (rms, switch), (drms, dswitch) = jax.jvp(self.compute_primal,\n",
+    "                                                     primals=(x,),\n",
+    "                                                     tangents=(tangents_x,))\n",
+    "            drms_dx = drms_dx.at[i].set(drms)\n",
+    "            dswitch_dx = dswitch_dx.at[i].set(dswitch)\n",
+    "            tangents_x = tangents_x.at[i].set(0.)\n",
+    "\n",
+    "        return drms_dx, dswitch_dx\n",
+    "\n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_vjp(self, x):\n",
+    "        # VJP is a good choice here since the jacbian matrix is a row vector.\n",
+    "        # We can compute the jacobian with a single call of the VJP function.\n",
+    "   \n",
+    "        # vjp always returns the primal and the vjp\n",
+    "        primal, vjp_fun = jax.vjp(self.compute_primal, x)\n",
+    "        \n",
+    "        # Get the partials drms_dx\n",
+    "        cotangents = (jnp.ones_like(primal[0]), jnp.zeros_like(primal[0]))\n",
```

Comment:
since jax arrays are immutable, could create temporaries for `jnp.zeros_lik(primal[0])` and `jnp.ones_like(primal[0])`

Reply:
I was hesitant to bother saving these off as attributes of self and then complicating the tree_flatten function in the very first example.  I'm not sure if it's worth it here.